flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java

```diff
@@ -0,0 +1,197 @@
+                oldState,
+                state);
+    }
+
+    @Override
+    public synchronized void handleExecutionTermination(
+            Collection<ExecutionState> executionStates) {
+        final StopWithSavepointState oldState = state;
+        state = state.onExecutionsTermination(this, executionStates);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on execution termination handling.",
+                oldState,
+                state);
+    }
+
+    @Override
+    public CompletableFuture<String> getResult() {
+        return result;
+    }
+
+    private StopWithSavepointState terminateExceptionWithGlobalFailover(
+            Iterable<ExecutionState> unfinishedExecutionStates) {
+        String errorMessage =
+                String.format(
+                        "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: %s. A global fail-over is triggered to recover the job %s.",
+                        StringUtils.join(unfinishedExecutionStates, ", "), jobId);
+        FlinkException inconsistentFinalStateException = new FlinkException(errorMessage);
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+        return terminateExceptionally(inconsistentFinalStateException);
+    }
+
+    private StopWithSavepointState terminateExceptionally(Throwable throwable) {
+        scheduler.startCheckpointScheduler(checkpointCoordinator);
+        result.completeExceptionally(throwable);
+
+        return StopWithSavepointState.Final;
+    }
+
+    private StopWithSavepointState terminateSuccessfully(String path) {
+        result.complete(path);
+
+        return StopWithSavepointState.Final;
+    }
+
+    private static Set<ExecutionState> extractUnfinishedStates(
+            Collection<ExecutionState> executionStates) {
+        return executionStates.stream()
+                .filter(state -> state != ExecutionState.FINISHED)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * {@code StopWithSavepointState} represents the different states during the stop-with-savepoint
+     * operation.
+     *
+     * <p>The state transitions are implemented in the following way: InitialWait ->
+     * [WaitForSavepointCreation|WaitForJobTermination] -> Final
+     */
+    private enum StopWithSavepointState {
+        InitialWait {
+            @Override
+            protected StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                context.path = path;
+                return WaitForJobTermination;
+            }
+
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                context.unfinishedStates = extractUnfinishedStates(executionStates);
```

Comment:
Couldn't we already trigger a global failover if `unfinishedStates.isEmpty() == false`?

Reply:
The code was refactored making this comment obsolete.